src/models/priority_queue.ts

```diff
@@ -0,0 +1,185 @@
+/**
+ * Copyright (c) Hathor Labs and its affiliates.
+ *
+ * This source code is licensed under the MIT license found in the
+ * LICENSE file in the root directory of this source tree.
+ */
+/* eslint class-methods-use-this: ["error", { "exceptMethods": ["parent", "left", "right", "comparator"] }] */
+
+interface PriorityQueueNode<T> {
+  value: T;
+  priority: number;
+}
+
+/**
+ * Priority queue implementation using an underlying heap for performance.
+ * @example
+ * const prioQ = new PriorityQueue<string>();
+ * prioQ.push(prioQ.makeNode(0, 'lower prio'));
+ * prioQ.push(prioQ.makeNode(10, 'higher prio'));
+ * prioQ.push(prioQ.makeNode(5, 'medium prio'));
+ * // Returns 'higher prio'
+ * prioQ.pop()
+ * // Returns 'medium prio'
+ * prioQ.pop()
+ * // Returns 'lower prio'
+ * prioQ.pop()
+ *
+ * @template [T=unknown]
+ */
+export default class PriorityQueue<T = unknown> {
+  #heap: PriorityQueueNode<T>[];
+
+  readonly #top: number;
+
+  constructor() {
+    this.#heap = [];
+    this.#top = 0;
+  }
+
+  /**
+   * Utility to create a node from a value and priority.
+   * @example
+   * // returns { value: 'foobar', 'priority': 10 }
+   * prioQueue.makeNode(10, 'foobar');
+   */
+  static makeNode<N>(priority: number, value: N): PriorityQueueNode<N> {
```

Comment:
If I use `T` here it is not a static method anymore because we use `T` is the generic type of an instance.
If the method is static it should provide functionality for all instances, hence a new generic.

Reply:
I see, makes sense